hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/MemberSchemaService.java

```diff
@@ -0,0 +1,188 @@
+        Schema schema = getLocal(schemaId);
+        if (schema != null) {
+            return CompletableFuture.completedFuture(schema);
+        }
+        if (logger.isFinestEnabled()) {
+            logger.finest("Could not find schema id  " + schemaId + " locally, will search on the cluster" + schemaId);
+        }
+        ClusterService cluster = nodeEngine.getClusterService();
+        OperationService operationService = nodeEngine.getOperationService();
+        Set<Member> members = cluster.getMembers();
+        Iterator<Member> iterator = members.iterator();
+        return searchClusterAsync(schemaId, iterator, operationService);
+    }
+
+    private CompletableFuture<Schema> searchClusterAsync(long schemaId, Iterator<Member> iterator,
+                                                         OperationService operationService) {
+        if (!iterator.hasNext()) {
+            return CompletableFuture.completedFuture(null);
+        }
+        Address address = iterator.next().getAddress();
+        FetchSchemaOperation op = new FetchSchemaOperation(schemaId);
+        InvocationFuture<Schema> future = operationService.invokeOnTarget(SERVICE_NAME, op, address);
+        return future.handle((data, throwable) -> {
+            //handle the exception and carry it to next `thenCompose` method
+            if (throwable != null) {
+                return throwable;
+            }
+            return data;
+        }).thenCompose(o -> {
+            if (o instanceof Throwable || o == null) {
+                return searchClusterAsync(schemaId, iterator, operationService);
+            }
+            Schema retrievedSchema = (Schema) o;
+            putLocal(retrievedSchema);
+            return CompletableFuture.completedFuture(getLocal(schemaId));
+        });
+    }
+
+    public Schema getLocal(long schemaId) {
+        return schemas.get(schemaId);
+    }
+
+    @Override
+    public void put(Schema schema) {
+        if (!seenSchemas.add(schema)) {
+            //this is to prevent converting every schema put to data and calculate fingerprint
+            return;
+        }
+        long schemaId = schema.getSchemaId();
+        if (putIfAbsent(schema)) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Sending schema id  " + schemaId + " locally, will search on the cluster" + schemaId);
+            }
+            if (nodeEngine.getClusterService().getClusterState().equals(ClusterState.PASSIVE)) {
+                return;
+            }
+            invokeOnStableClusterSerial(nodeEngine, () -> new SendSchemaOperation(schema), MAX_RETRIES)
+                    .joinInternal();
+        }
+    }
+
+    public CompletableFuture<Void> putAsync(Schema schema) {
+        if (putIfAbsent(schema)) {
+            if (nodeEngine.getClusterService().getClusterState().equals(ClusterState.PASSIVE)) {
+                return CompletableFuture.completedFuture(null);
```

Comment:
This code was introduced to load the data from HotRestart where clusterState is passive. 
This code should not be needed anymore since we are putting the schema to local first on trimSchema.

Reply:
Fixed by d4d741e